language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+
+use crate::logging::AdapterLogSchema;
+
+/// This trait describes the VM adapter's interface.
+pub trait VMAdapter {
+    /// Creates a new Session backed by the given storage.
+    /// TODO: this probably doesn't belong in this trait. We should be able to remove
+    /// this after refactoring Diem transaction executor code.
+    fn new_session<'r, R: MoveStorage>(&self, remote: &'r R) -> Session<'r, '_, R>;
+
+    /// Checks the signature of the given signed transaction and returns
+    /// `Ok(SignatureCheckedTransaction)` if the signature is valid.
+    fn check_signature(txn: SignedTransaction) -> Result<SignatureCheckedTransaction>;
+
+    /// Returns true if multi agent transactions are allowed.
+    fn allows_multi_agent(&self) -> Result<bool, VMStatus>;
+
+    /// Runs the prologue for the given transaction.
+    fn run_prologue<S: MoveStorage>(
+        &self,
+        session: &mut Session<S>,
+        transaction: &SignatureCheckedTransaction,
+        currency_code: &Identifier,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(), VMStatus>;
+}
+
+/// Validate a signed transaction by performing the following:
+/// 1. Check the signature(s) included in the signed transaction
+/// 2. Check that the transaction is allowed in the context provided by the `adapter`
+/// 3. Run the prologue to perform additional on-chain checks
+/// The returned `VMValidatorResult` will have status `None` and if all checks succeeded
+/// and `Some(DiscardedVMStatus)` otherwise.
+pub fn validate_signed_transaction<A: VMAdapter>(
+    adapter: &A,
+    transaction: SignedTransaction,
+    state_view: &dyn StateView,
+) -> VMValidatorResult {
+    let _timer = TXN_VALIDATION_SECONDS.start_timer();
+    let txn_sender = transaction.sender();
+    let log_context = AdapterLogSchema::new(state_view.id(), 0);
+
+    let txn = if let Ok(t) = A::check_signature(transaction) {
+        t
+    } else {
+        return VMValidatorResult::error(StatusCode::INVALID_SIGNATURE);
+    };
+
+    let remote_cache = StateViewCache::new(state_view);
+    let account_role = get_account_role(txn_sender, &remote_cache);
+    let mut session = adapter.new_session(&remote_cache);
+
+    let (status, normalized_gas_price) = match validate_signature_checked_transaction(
+        adapter,
+        &mut session,
+        &txn,
+        &remote_cache,
+        true,
+        &log_context,
+    ) {
+        Ok((price, _)) => (None, price),
+        Err(err) => (Some(err.status_code()), 0),
+    };
+
+    // Increment the counter for transactions verified.
+    let counter_label = match status {
+        None => "success",
+        Some(_) => "failure",
+    };
+    TRANSACTIONS_VALIDATED
+        .with_label_values(&[counter_label])
+        .inc();
+
+    VMValidatorResult::new(status, normalized_gas_price, account_role)
+}
+
+fn get_account_role(sender: AccountAddress, remote_cache: &StateViewCache) -> GovernanceRole {
```

Comment:
This is also Diem-specific. I think this is here because the Diem mempool looks at the role to decide how to prioritize transactions: sort first by role, then by gas price. If we refactor the mempool/adapter API to return an `impl PartialOrd` (or similar) instead, I think we can keep the role stuff inside `DiemVM` and eliminate this from the public API.

Reply:
Makes sense. Will look into this.

Reply:
I looked into how mempool is using this information right now. It seems that we’ll have to change the code here to be generic: https://github.com/diem/diem/blob/main/mempool/src/core_mempool/index.rs#L54. Will work on this in a future PR.